includes/class-wc-gateway-payfast.php

```diff
@@ -1796,4 +1797,45 @@ public function display_order_net( $order_id ) {
 
 		<?php
 	}
+
+	/**
+	 * Filters the currency to 'ZAR' if set via WooPayments multi-currency feature.
+	 *
+	 * @param string $currency The currency code.
+	 * @return string
+	 */
+	public function filter_currency( $currency ) {
+		// Do nothing if WooPayments is not activated.
+		if ( ! class_exists( '\WCPay\MultiCurrency\MultiCurrency' ) ) {
+			return $currency;
+		}
+
+		// Do nothing if the page is admin screen.
+		if ( is_admin() ) {
+			return $currency;
+		}
+
+		$user_id = get_current_user_id();
+
+		// Check if the currency is set in the URL.
+		if ( isset( $_GET['currency'] ) ) { // phpcs:ignore WordPress.Security.NonceVerification.Recommended
+			$currency_code = array(
+				sanitize_text_field(
+					wp_unslash( $_GET['currency'] ) // phpcs:ignore WordPress.Security.NonceVerification.Recommended
+				),
+			);
+			// Check if the currency is set in the session (for logged-out users).
+		} elseif ( 0 === $user_id && WC()->session ) {
+			$currency_code = WC()->session->get( \WCPay\MultiCurrency\MultiCurrency::CURRENCY_SESSION_KEY );
+			// Check if the currency is set in the user meta (for logged-in users).
+		} elseif ( $user_id ) {
+			$currency_code = get_user_meta( $user_id, \WCPay\MultiCurrency\MultiCurrency::CURRENCY_META_KEY );
+		}
```

Comment:
I'm not familiar with how multi-currency works so this feedback may not make sense. But wondering if there's a scenario where the user session or user meta hasn't been set and thus the currency code ends up being empty here?

I guess wondering if we should separate this code into two `if` statements. So we check for the currency value from the URL first and store that in the `$currency_code` variable. And then we overwrite that value from session or user meta, if that value exists in either. Something like:
```php
$currency_code = '';
if ( isset( $_GET['currency'] ) ) { // phpcs:ignore WordPress.Security.NonceVerification.Recommended
	$currency_code = array(
		sanitize_text_field(
			wp_unslash( $_GET['currency'] ) // phpcs:ignore WordPress.Security.NonceVerification.Recommended
		),
	);
}

// Check if the currency is set in the session (for logged-out users).
if ( 0 === $user_id && WC()->session ) {
	$stored_code = WC()->session->get( \WCPay\MultiCurrency\MultiCurrency::CURRENCY_SESSION_KEY );

	// Not sure the best check on this value? Should it always be an array and we can check for that? Can it sometimes be a string?
	if ( $stored_code ) {
		$currency_code = $stored_code;
	}
// Check if the currency is set in the user meta (for logged-in users).
} elseif ( $user_id ) {
	$stored_code = get_user_meta( $user_id, \WCPay\MultiCurrency\MultiCurrency::CURRENCY_META_KEY );

	// Not sure the best check on this value? Should it always be an array and we can check for that? Can it sometimes be a string?
	if ( $stored_code ) {
		$currency_code = $stored_code;
	}
}
```

Reply:
@dkotter I've fixed the confusion between the return value types `array` vs `string`. There was a small bug, thanks for pointing that out.

Regarding separating if statements into 2, the problem is that in Block Checkout, the user meta is saved after the block is rendered. So the block checkout page requires an additional reload for Payfast to appear in the list of gateways. We give precedence to the GET parameter, if that is set, then we use it, because ultimately that gets saved to the session and user meta.